icu4c/source/common/locdspnm.cpp

```diff
@@ -575,28 +546,37 @@ LocaleDisplayNamesImpl::localeDisplayName(const Locale& loc,
   UBool hasVariant = uprv_strlen(variant) > 0;
 
   if (dialectHandling == ULDN_DIALECT_NAMES) {
-    char buffer[ULOC_FULLNAME_CAPACITY];
+    UErrorCode status = U_ZERO_ERROR;
+    CharString buffer;
     do { // loop construct is so we can break early out of search
       if (hasScript && hasCountry) {
-        ncat(buffer, ULOC_FULLNAME_CAPACITY, lang, "_", script, "_", country, (char *)0);
-        localeIdName(buffer, resultName, false);
+        buffer.append(lang, status);
+        buffer.append('_', status);
+        buffer.append(script, status);
+        buffer.append('_', status);
+        buffer.append(country, status);
```

Comment:
optional: we can use function chaining
```suggestion
        buffer.append(lang, status).append('_', status).append(script, status).append('_', status).append(country, status);
```